core/domain/html_cleaner.py

```diff
@@ -405,7 +450,38 @@ def convert_to_text_angular(html_data):
     # html strings they are stored as <br>. Since both of these
     # should match and <br> and <br/> have same working,
     # so the tag has to be replaced in this way.
-    return str(soup).replace('<br/>', '<br>')
+    return unicode(soup).replace('<br/>', '<br>')
+
+
+def tag_content_to_textangular(html_data):
+    """This functions converts content with tabs and collapsible
+    components to textangular format.
+
+    Args:
+        html_data: str. The HTML string whose content is to be converted.
+
+    Returns:
+        str. The HTML string with converted content within tag.
+    """
+    soup = bs4.BeautifulSoup(html_data.encode('utf-8'), 'html.parser')
+
+    for collapsible in soup.findAll('oppia-noninteractive-collapsible'):
+        content_html = html_unescaper(collapsible['content-with-value'])
+        # This is done to ensure that &quot; is not escaped as &quot;
+        # is not escaped as &amp;amp;quot;.
+        collapsible['content-with-value'] = '&quot;' + html_escaper(
```

Comment:
This smells fishy. Why do you need to do it here and not for tabs? It's unclear to me exactly what the issue is and why you need to do this at all. This also assumes that the first and last elements of content_html are quotes -- this is an assumption that must be checked, not just assumed, if you go down this path.

Reply:
This is because collapsible and tabs contain RTE content (stored in an attribute), and that content also needs to be migrated. I agree there needs to be a case for tabs too (tabs stores RTE content under `tab_content-with-value`).

Looking at `before_migration.yaml`, it does seem the contents of `content-with-value` is always surrounded by quotes and ampersands (`&amp`). Nitish: Do you understand what part of the escaping introduces these quotes and ampersands? Also, why only add the quotes (and not the ampersands)?

Reply:
Hi @AllanYangZhou @seanlip 

In collapsible the content is as follows:

``<oppia-noninteractive-collapsible content-with-value=\"&amp;quot;Hello oppia&amp;quot;\" heading-with-value=\"&amp;quot;Learn more about APIs&amp;quot;\"></oppia-noninteractive-collapsible>``
So after unescaping, the content-with-value is ``"Hello oppia"`` i.e. quotes are a part of html string, so they get wrapped inside p tag which is not what we want. To remove the quotes, I extract the content leaving the first and last character which are quotes and add them back (after performing migration) in content-with-value.

In tabs the content is as follows:
``<oppia-noninteractive-tabs tab_contents-with-value=\"[{&amp;quot;content&amp;quot;:&amp;quot;&amp;lt;span style=\\&amp;quot;line-height: 21px; background-color: rgb(255, 255, 255);\\&amp;quot;&amp;gt;lorem ipsum&amp;lt;/span&amp;gt;&amp;quot;,&amp;quot;title&amp;quot;:&amp;quot;hello&amp;quot;}]\"></oppia-noninteractive-tabs>``

Here the content inside tabs_content-with-value on unescaping will not have quotes as part of html string (since this is stored as dict: key-value pair). So, no special handling is needed as in the previous case.

Also to answer @AllanYangZhou 's question `&amp;' is not necessary along with `quot;`. Escaping only to `&quot;' also works fine (since '&amp;' is converted to '&').

Thanks

Reply:
To expand on why `content-with-value` (or other `-with-value` attrs) always start and end with `&quot`:

In the RTE, data is stored into a `-with-value` attr using `objToEscapedJson`, which first calls `JSON.stringify`, which introduces the surrounding quotes (try calling `JSON.stringify` on an arbitrary JS string). Escaping then convertes the surrounding quotes to `&quot;`.

Reply:
So I still think we should avoid arbitrary constructions like [1:-1], especially if we don't do something equivalent when constructing the attribute value in the first place.

Couldn't we reverse the construction process in a generic way? @AllanYangZhou mentions that, to get from data to the attr, we do json.dumps() and then escape. So here we should unescape and then do json.loads(), right?